### DISCUSSION TITLE
How would you do it? mpremote web browser

### DISCUSSION THREAD

**Discussion by patfrench** [micropython/micropython#12796] [General]
I have several questions about my project: 

- [x] I have a RPI-PICO built with the WEBUSB version. [https://github.com/micropython/micropython/pull/11301](https://github.com/micropython/micropython/pull/11301)

- [x] I've developed a javascript code that connects to the RPI-PICO via USB and gives me access to the REPL. Tested with
[https://example.tinyusb.org/webusb-serial/](https://example.tinyusb.org/webusb-serial/)

- [x] I've added "pyscript" to the javascript, so I now have a python interpreter in my browser.

- [x] I've installed mpremote in my browser's python environment
[https://titeppat.pyscriptapps.com/tight-bonus/latest/](https://titeppat.pyscriptapps.com/tight-bonus/latest/)

- [ ] I'd like to use mpremote but I'm lost. 
- [ ] I need to replace pyserial with my USB connection ?
- [ ] I need to be able to use mpremote in the browser interpreter ? 

is this possible? Do you have any suggestions ?

Thank you

**Comment by andrewleech** [comment]
If you look at /tools/mpremote in this repo the current version might be easier to follow, the transport_serial file handles all the communicating.

I'm not sure about python in the browser though, whether it's possible to get read/write stream date out to a JavaScript function which could then forward it on to the webusb functions.

**Comment by patfrench** [comment]
Ok

I saw that Jeff Glass did the same thing with the WEBSERIAL API. [https://jeff.glass/post/pyscript-webserial/](https://jeff.glass/post/pyscript-webserial/)
I'll have a look

I need to know if i can use mpremote as a library ? can i do like the pyboard.py utility ?
```import pyboard
pyb = pyboard.Pyboard('/dev/ttyACM0', 115200)
pyb.enter_raw_repl()
ret = pyb.exec('print(1+1)')
print(ret)
pyb.exit_raw_repl()```

thank you

**Reply by Josverl** [reply]
Mpremote currently does not have a well defined api, there is a PR with a partial implementation

https://github.com/micropython/micropython/pull/11777

**Comment by patfrench** [comment]
Hi,

Finally, I write the functions in javascript. I don't have enough skills to transform mpremote or pyboard.py 

I still need your help:

The version of micropython [https://github.com/KitronikLtd/micropython-rp2-webusb](https://github.com/KitronikLtd/micropython-rp2-webusb) is 1.19

And I use a PICO_W, **I don't know how difficult it is to recompile firmware to version 1.20**. 

Because I would like to have access to "import bluetooth ".

Thank you

**Comment by vshymanskyy** [comment]
Check out `MicroPython Web IDE`: https://github.com/orgs/micropython/discussions/15219

**Reply by patfrench** [reply]
Hi,
Very good job.
This is my web editor (webUSB ans web bluetooth)
https://github.com/patfrench/patfrench.github.io

Test démo page :
https://patfrench.github.io/

Thank you

**Reply by vshymanskyy** [reply]
Regarding BLE, which devices do you usually test with? Do you need to run a special firmware for this?

**Reply by patfrench** [reply]
hi,

for the BLE, i use : [https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_uart_repl.py](https://github.com/micropython/micropython/blob/master/examples/bluetooth/ble_uart_repl.py)

In my file system there is : 
![image](https://github.com/micropython/micropython/assets/44723412/3e359351-e0be-4532-8235-bba342937137)

and in my main.py (maybe in boot.py it would have been better)
```python
import ble_uart_repl
```

thank you

**Reply by patfrench** [reply]
ble_uart_repl will work with original firmware.

for webusb no, it's a custome firmware, i don't use webserial.

for your android version: how does it work? with webserial?
